e2e/loadtest/load_test.go

```diff
@@ -0,0 +1,146 @@
+	opts.Duration = time.Duration(*duration) * time.Second
+
+	// Run memory profiling to find out memory stability of SDK
+	err := testutil.StartEnvoyMemoryProfile(8001)
+	require.NoError(t, err)
+
+	// Start generating load
+	results, err := fhttp.RunHTTPTest(&opts)
+	require.NoError(t, err)
+
+	// Stop memory profiling
+	memstats, err := testutil.StopEnvoyMemoryProfile()
+	require.NoError(t, err)
+
+	require.GreaterOrEqualf(t, results.ActualQPS, *qps, "Actual QPS should be higher than target QPS")
+
+	// TODO(musaprg): Currently, we're observing memory usage of envoy for checking memory stability,
+	// but we need to use customized tinygo which enable to hook the timing of GC execution for finding
+	// a best latency-friendly timing to invoke GC. Currently, tinygo invokes GC when memory allocations
+	// attempts. It might affect the response latency. Basically, we don't need to invoke GC manually,
+	// but it's better to do it out of the request-processing time.
+
+	// Summarizing memory profile
+	heapSizes := []float64{}
+	allocSizes := []float64{}
+	maxUsage := float64(0)
+	maxAllocSize := float64(0)
+	maxIndex := 0
+	for i, m := range memstats {
+		heapUsage := float64(m.Allocated) / float64(m.HeapSize)
+		allocSize := float64(m.Allocated)
+		heapSize := float64(m.HeapSize)
+		allocSizes = append(allocSizes, allocSize)
+		heapSizes = append(heapSizes, heapSize)
+		if maxUsage < heapUsage {
+			maxUsage = heapUsage
+			maxIndex = i
+		}
+		if maxAllocSize < allocSize {
+			maxAllocSize = allocSize
+		}
+	}
+	log.Printf("peak memory usage: %v (elapsed %f sec after invoking load test)", maxUsage, float64(maxIndex*100)/1000)
+	log.Printf("peak memory: %d bytes (+%d bytes increased from beginning)", int64(maxAllocSize), int64(maxAllocSize-allocSizes[0]))
+
+	// Plotting memory profile
+	p := plot.New()
+	p.Title.Text = fmt.Sprintf("Heap profiling of envoy process (%f QPS, %s)", *qps, *targetExample)
+	p.X.Label.Text = "elapsed time [ms]"
+	p.Y.Label.Text = "memory size [KB]"
+	heapSizePlot := make(plotter.XYs, len(heapSizes))
+	for i, v := range heapSizes {
+		heapSizePlot[i].X = float64(i * 100)
+		heapSizePlot[i].Y = v / 1024 // Convert to KB
+	}
+	allocSizePlot := make(plotter.XYs, len(allocSizes))
+	for i, v := range allocSizes {
+		allocSizePlot[i].X = float64(i * 100)
+		allocSizePlot[i].Y = v / 1024 // Convert to KB
+	}
+	err = plotutil.AddLinePoints(p,
+		"heap_size", heapSizePlot,
+		"allocated", allocSizePlot)
+	require.NoError(t, err)
+
+	// Save the plot to a PNG file.
+	err = p.Save(font.Length(len(heapSizes))*vg.Millimeter, 8*vg.Inch, "/tmp/memstat.png")
+
+	fortioLog.WriteTo(log.Writer())
+
+	successRate := float64(results.RetCodes[200]) / float64(results.DurationHistogram.Count)
+	require.GreaterOrEqual(t, successRate, targetSuccessRate, stdErr.String())
+	require.LessOrEqual(t, results.DurationHistogram.Percentiles[0].Value, float64(targetNintyninthPercentileLatencyLimit), stdErr.String())
```

Comment:
I think we have to also check not only 99%, but also max - this is how we can see the impacts of GC.